### ISSUE TITLE
Sigmoid Surrogate calculating NaNs on backward pass

### ISSUE THREAD

**Issue by MeinChef** [jeshraghian/snntorch#427]
**Issue type**

- [x] Bug Report
- [ ] Feature Request
- [ ] Help wanted
- [ ] Other

**snnTorch version**

`0.9.4`

**Operating System**

Reproduceable on Fedora `7.0.11-200.fc44.x86_64` and CachyOS `7.0.10-2-cachyos`

**Description**

I first stumbled upon this when I was training my model, and could not explain weird loss-behaviours.

When using the [Sigmoid Surrogate](https://snntorch.readthedocs.io/en/latest/snntorch.surrogate.html#snntorch.surrogate.Sigmoid) with the default slope of 25, some gradients are NaN. 

This is probably due to the exponentiation in the surrogate not being numerically stable. 
In the source code, on the `backward()` method, [one line](https://github.com/jeshraghian/snntorch/blob/634fc22385ec5581998e62787bbe39b53662140f/snntorch/surrogate.py#L307) is:
```python
torch.exp(-ctx.slope * input_)
```
with e.g. an input of `-5`, $exp(125)$ is being calculated, which is mathematically sound, but for torch this is `inf`.

**Minimal code to reproduce the error/bug**

```python
import torch
import snntorch.surrogate as surr
torch.autograd.set_detect_anomaly(True)
x = torch.tensor([-5.0], requires_grad = True)
y = surr.sigmoid()(x)
y.backward()
```
Or alternatively used in a network with dummy spike inputs:
```python
import snntorch as snn
import snntorch.surrogate as surr
import snntorch.functional as sf
import torch

class Model(torch.nn.Module):
    def __init__(self) -> None:
        super().__init__()

        # hidden layer
        self._in = torch.nn.Linear(
            in_features = 8,
            out_features = 16,
        )
        self._hidden = snn.Leaky(
            beta = 0.9,
            spike_grad = surr.sigmoid()
        )

        # out
        self._con = torch.nn.Linear(
            in_features = 16,
            out_features = 2
        )
        self._out = snn.Leaky(
            beta = 0.9,
            spike_grad = surr.sigmoid()
        )

    def forward(
        self,
        x: torch.Tensor
    ) -> torch.Tensor:
        mem1 = self._hidden.reset_mem()
        mem2 = self._out.reset_mem()

        out = []
        for step in range(x.shape[0]):
            # hidden layer
            cur1 = self._in(x[step])
            spk1, mem1 = self._hidden(cur1, mem1)

            # out layer
            cur2 = self._con(spk1)
            spk2, mem2 = self._out(cur2, mem2)

            out.append(spk2)
        
        return torch.stack(out)


if __name__ == "__main__":
    # init stuff
    if torch.cuda.is_available():
        device = torch.device("cuda")
    elif torch.backends.mps.is_available():
        device = torch.device("mps")
    else: 
        device = torch.device("cpu")
    
    model = Model()
    model.to(device)
    lossfn = sf.loss.mse_count_loss()

    # create a bit of dummy data of shape time_steps, batch_size, in_features
    data = torch.randint(
        low = 0,
        high = 2,
        size = (6,2,8), 
        dtype = torch.float32,
        device = device
    )
    targets = torch.randint(
        low = 0,
        high = 2, # two classes
        size = (2,),
        device = device
    ) 

    # training
    model.train()
    pred = model(data)
    
    # torch.autograd.set_detect_anomaly(True)
    loss = lossfn(pred, targets) 
    loss.backward()

    # print the stuff that's relevant
    for param in model.parameters():
        if param.grad is not None:
            nans = param.grad.isnan()
            print(
                f"Found {nans.sum()} NaNs in the Gradients.\n"
                f"Shape of corresponding layer: {param.grad.shape}"
            )
```

**Proposal**

I'm not the smartest of people, but I remember  a derivative of the sigmoid function should be $\sigma'(x) = \sigma(x)\left(1-\sigma(x)\right)$.
Using this and the `torch.sigmoid()` function, which should be numerically stable™, one could rewrite the function as follows:
```python
@staticmethod
def backward(ctx, grad_output):
    (input_,) = ctx.saved_tensors
    grad_input = grad_output.clone()
    sig = torch.sigmoid(ctx.slope * input_)
    grad = (
      grad_input
      * ctx.slope
      * sig
      * (1 - sig)
    )
    return grad, None
```

**Comment by ixfd64** [comment]
Should be fixed in #433. Feel free to re-open if there are still issues.
